### ISSUE TITLE
Thead::Row::Cell is not rendered in v2

### ISSUE THREAD

**Issue by jelhan** [miguelcobain/ember-yeti-table#464]
There is a timing error, which breaks usage of `<Thead::Row::Cell>`. The cell is not rendered.

This is a simplified use case, which I'm having:

```hbs
<YetiTable
    @data={{array (hash firstName="Max" lastName="Mustermann") (hash firstName="John" lastName="Doe")}}
    as |table|
>
    <table.thead as |thead|>
        <thead.row as |row|>
            <row.column @prop="firstName">
                first name
            </row.column>
            <row.column @prop="lastName">
                last name
            </row.column>
        </thead.row>
        <thead.row as |row|>
            <row.cell>
                <input placeholder="first name">
            </row.cell>
            <row.cell>
                <input placeholder="last name">
            </row.cell>
        </thead.row>
    </table.thead>
    <table.body />
</YetiTable>
```

The second `<thead.row>`, which uses `<row.cell>` is not rendered.

I debugged it and noticed that `this.args.columns` is an empty array when the cell is registered here: https://github.com/miguelcobain/ember-yeti-table/blob/216485a33c8c0aff15b75f65f3dec494facd718c/addon/components/yeti-table/thead/row.gjs#L42-L58

The column is registered here: https://github.com/miguelcobain/ember-yeti-table/blob/216485a33c8c0aff15b75f65f3dec494facd718c/addon/components/yeti-table.gjs#L703-L713 That function is executed _before_ `registerCell`. But the work is scheduled for `afterRender`, which is executed _after_ `registerCell`.

Please find a reproduction here: https://github.com/jelhan/reproduction-ember-yeti-table-cell-registration-issue

**Comment by cah-brian-gantzler** [comment]
V2 went from components and 2 way binding to glimmer components and DDAU. In many ways a lot of the control and flow was rewritten. 

Im wondering why registerColumn is scheduled afterRender (or scheduled at all). Would have to look at that. 

Currently there is a PR out https://github.com/miguelcobain/ember-yeti-table/pull/462 that we should get merged first. Can take a look at this after that. Thanks for the report.

**Comment by jelhan** [comment]
If we can remove the scheduling after render that would resolve the issue. Likely the best option to investigate first.

Also noticed that there were many changes between v1 and v2. Therefore I have up tracking it down to a specific commit.

**Comment by jelhan** [comment]
I tested if we can remove scheduling column registration on `afterRender`. But this leads to modification after consumption errors. Doesn't seem to be an easy fix. :cry: 

I pushed the test, which I have written for the experiment in #465. May help for others testing out potential fixes as well.

**Comment by cah-brian-gantzler** [comment]
Thanks for confirming the reason the schedule is there. That particular error can be annoying to fix. I have that issue right now. Updating from ember 5.5 to 5.6 causes that issue in 3 tests for my application that I have to hunt down.

**Comment by jelhan** [comment]
Yeah. That indicates an architectural issue, which was worked around by scheduling outside of rendering loop.

If we are lucky, the only consumption is directly before modification: https://github.com/miguelcobain/ember-yeti-table/blob/216485a33c8c0aff15b75f65f3dec494facd718c/addon/components/yeti-table.gjs#L709-L711

Maybe we can convert `columns` to a `TrackedSet`. In that case we wouldn't need to check if it's already in the list.

As a next step we could check if `columns` actually need to be tracked.

**Comment by jelhan** [comment]
I tried using a `TrackedSet` but also run into the update after usage error.

Main implementation looked like this:

```patch
   registerColumn(column) {
-    schedule('afterRender', this, function () {
-      if (typeof this.args.isColumnVisible === 'function') {
-        column.visible = this.args.isColumnVisible(column);
-      }
+    if (typeof this.args.isColumnVisible === 'function') {
+      column.visible = this.args.isColumnVisible(column);
+    }
 
-      if (!this.columns.includes(column)) {
-        this.columns = [...this.columns, column];
-      }
-    });
+    this.columns.add(column);
   }
```

Update after usage error was thrown by `this.columns.add(column)`.

**Comment by cah-brian-gantzler** [comment]
The whole reason for the cell having a column property is for visible. So if the column is hidden, the cell is hidden. 

Since we determined that the afterRender has to happen for columns, First guess is that the cell register should also be after render for the same reason. I think that this will fix the problem for the common situation where the first row is column and the second row is cell.

If the first row is cell and the second row is column, this will not work. I also have concerns when there is a colspan involved on the column or the cell. Several variations and it will not work for any column/cell other than index 0 for the cell/column. There should be an issue devoted to how to use colspan when having two headers. 

The column for the cell is looked up at creation. If the lookup was delayed until the `if this.column.visible` is evaluated, it would also probably solve the problem and be a lot more reactive to changes on the column. 

I would suggest making the if hide the column reference and the template for the cell this
```
<template>
    {{#if this.isVisible}}
      <th class='{{@class}} {{@theme.theadCell}}' ...attributes>
        {{yield}}
      </th>
    {{/if}}
  </template>
```
  
remove the column definition
```
// Assigned when the cell is registered
  column = undefined;
```
and add 
```javascript
  get isVisible() {
    const props = Array.isArray(this.prop) ? this.prop : [this.prop];
    return props.map(prop => this.args.columns.find((col) => col.prop === prop)?.visible).every((visible) => visible);
  }
```
Now you dont need to register cell at all, can remove the call and the willDestroy on the cell. You can also remove the registerCell method and array on row.

What is all the extra code about you ask.     Well, if there is a colspan at all on either of the header rows, the whole code fails miserably unless it is the first column only. 

When you have two rows, you MUST have a prop on the column and cell to link them even if there is no colspan. It would be a mess to support no prop when there are no colspans, so just make it mandatory. There may be one column and two+ cells (cells will use the same prop), or one cell and two+ columns, the prop on cell will be `{{array 'prop1' 'prop2'}}`. Only when all linked columns are not visible, will the cell not be visible. 

The double header row never supported colspan correctly, now in addition to fixing your problem, it actually supports colspans as well. 

I have not tested this code yet until I can get tests to run. Please forgive any typos. 

NOTE 1: Original code used findBy, these are removed in ember 6.0.   Need to replace filterBy, findBy, etc. That is why I used filter and map in the above code.

**Comment by jelhan** [comment]
@cah-brian-gantzler Thanks a lot for your detailed investigation! I fully missed looking at _why_ we the mapping of cells to columns exists at all. After you mentioned that it is for synchronizing the visibility I noticed that there is the same requirement for body and footer cells. But that those are not affected by the same bug. Copying the implementation for body and footer cells to header cells fixes it. Please see #465 

It does not address the issue with colspan raised by you. But colspan is also not supported yet by body and footer cells. I feel we should address it for all of them in a follow-up.
